File: tools/richos-service/companion-macos/Tests/RichOSCompanionCoreTests/MixDecisionTests.swift
import XCTest
@testable import RichOSCompanionCore

/// The mic-vs-tap failover (§6.4): one source stalling must not freeze the recording — write the
/// live side, silence-fill the dead side, alarm. Never lose the live half; never go silent.
final class MixDecisionTests: XCTestCase {
    let threshold: Int64 = 15_000

    func testBothLiveTakesCommonMinimum() {
        let d = MixDecision.decide(micAvail: 100, systemAvail: 60, micSilentMs: 0, systemSilentMs: 0,
                                   starveThresholdMs: threshold)
        XCTAssertEqual(d, MixDecision(frames: 60, useMic: true, useSystem: true,
                                      micStarved: false, systemStarved: false))
    }

    func testSystemStalledFailsOverToMicOnly() {
        // system silent past threshold, mic still delivering
        let d = MixDecision.decide(micAvail: 80, systemAvail: 0, micSilentMs: 0, systemSilentMs: 20_000,
                                   starveThresholdMs: threshold)
        XCTAssertEqual(d.frames, 80)
        XCTAssertTrue(d.useMic)
        XCTAssertFalse(d.useSystem)      // RIGHT silence-filled
        XCTAssertTrue(d.systemStarved)   // -> red alarm
        XCTAssertFalse(d.micStarved)
    }

    func testMicStalledFailsOverToSystemOnly() {
        let d = MixDecision.decide(micAvail: 0, systemAvail: 50, micSilentMs: 20_000, systemSilentMs: 0,
                                   starveThresholdMs: threshold)
        XCTAssertEqual(d.frames, 50)
        XCTAssertFalse(d.useMic)         // LEFT silence-filled
        XCTAssertTrue(d.useSystem)
        XCTAssertTrue(d.micStarved)
    }

    func testBothStalledWritesNothingButAlarmsBoth() {
        let d = MixDecision.decide(micAvail: 0, systemAvail: 0, micSilentMs: 20_000, systemSilentMs: 20_000,
                                   starveThresholdMs: threshold)
        XCTAssertEqual(d.frames, 0)
        XCTAssertTrue(d.micStarved)
        XCTAssertTrue(d.systemStarved)
    }

    func testBriefLagUnderThresholdDoesNotFailover() {
        // system briefly behind but within threshold -> stay aligned, wait (min = 0 this pump).
        let d = MixDecision.decide(micAvail: 40, systemAvail: 0, micSilentMs: 0, systemSilentMs: 5_000,
                                   starveThresholdMs: threshold)
        XCTAssertEqual(d.frames, 0)
        XCTAssertFalse(d.systemStarved)
    }
}

File: tools/richos-service/companion-windows/tests/RichOSCompanionCore.Tests/MixDecisionTests.cs
using RichOSCompanionCore;
using Xunit;

namespace RichOSCompanionCore.Tests
{
    /// <summary>
    /// The mic-vs-loopback failover (§6.4): a stalled source must never freeze the whole recording —
    /// write the live side, silence-fill the dead side, alarm. Peer of the macOS companion's
    /// <c>MixDecisionTests.swift</c>.
    /// </summary>
    public class MixDecisionTests
    {
        [Fact]
        public void BothLiveTakesCommonMinimum()
        {
            var d = MixDecision.Decide(micAvail: 100, systemAvail: 60, micSilentMs: 0, systemSilentMs: 0, starveThresholdMs: 15000);
            Assert.Equal(60, d.Frames);
            Assert.True(d.UseMic);
            Assert.True(d.UseSystem);
            Assert.False(d.MicStarved);
            Assert.False(d.SystemStarved);
        }

        [Fact]
        public void MicStarvedWritesSystemAndSilenceFillsLeft()
        {
            var d = MixDecision.Decide(micAvail: 0, systemAvail: 480, micSilentMs: 20000, systemSilentMs: 0, starveThresholdMs: 15000);
            Assert.Equal(480, d.Frames);
            Assert.False(d.UseMic);
            Assert.True(d.UseSystem);
            Assert.True(d.MicStarved);
            Assert.False(d.SystemStarved);
        }

        [Fact]
        public void SystemStarvedWritesMicAndSilenceFillsRight()
        {
            var d = MixDecision.Decide(micAvail: 480, systemAvail: 0, micSilentMs: 0, systemSilentMs: 20000, starveThresholdMs: 15000);
            Assert.Equal(480, d.Frames);
            Assert.True(d.UseMic);
            Assert.False(d.UseSystem);
            Assert.False(d.MicStarved);
            Assert.True(d.SystemStarved);
        }

        [Fact]
        public void BothStarvedWritesNothingButBothAlarm()
        {
            var d = MixDecision.Decide(micAvail: 0, systemAvail: 0, micSilentMs: 20000, systemSilentMs: 20000, starveThresholdMs: 15000);
            Assert.Equal(0, d.Frames);
            Assert.True(d.MicStarved);
            Assert.True(d.SystemStarved);
        }
    }
}
